Switch Float16 to LLVM's half

maleadt opened the pull request:
This PR builds on @vchuravy's many experiments switching to LLVM's `half` type for Float16 instead of our current `i16` software implementation, and the accompanying integration with (a Julia version of) compiler-rt to implement the necessary intrinsics.
My motivation for this is to improve GPU codegen, where the `i16` software implementation hurts us quite a bit.
Ref https://github.com/JuliaLang/julia/pull/26381 https://github.com/JuliaLang/julia/pull/18734 https://github.com/JuliaLang/julia/pull/18927 

In short:
- switch codegen of Float16 over to `half` from `i16`
- move the software implementation to a Runtime module
- use `@ccallable` to expose those functions under the expected intrinsic names

I went with a Julia runtime library, and not the Real Thing (tm), such that these changes are minimally invasive and just move the current software implementation around. We can always decide later to use LLVM's version. On OSX however, xcode already links LLVM's compiler-rt, so I disabled registration of our runtime functions there (alternatively, @vchuravy mentions we could customize the library names in the TLI so that we use the same software implementation across platforms).

TODO:
- verify PPC support, we might have to apply a patch: https://bugs.llvm.org/show_bug.cgi?id=39865

`range` tests fail locally due to this:

```julia
function muladd(x, y, z)
    x*y + z
end
muladd(Float16(0.5283), Float16(0.584), Float16(-6.294e-5))
# Float16(0.3086) on master
# Float16(0.3083) on this PR
```

`0.3083` is more accurate (`Float16(0.5283 * 0.584 + -6.294e-5) == Float16(0.3083)`), but I'd think that shouldn't happen as the result differs from computing incrementally:

```
julia> Float16(0.5283) * Float16(0.584)
Float16(0.3086)

julia> ans + Float16(-6.294e-5)
Float16(0.3086)
```

I don't see any `muladd` combination happening at the LLVM level, but the generated assembly performs the arithmetic in single instead of half precision (`vmulss` + `vaddss`).

Thoughts? Is this legal, and should the tests be fixed? @vchuravy mentions that the same happens with FP32 operations benefiting from x87's extended 80-bit precision, and the same reasoning could be applied here.

vchuravy committed:
Emit Float16 as half instead of i16.

vchuravy committed:
Move Float16 software implementation to a Julia runtime library.

maleadt commented:
Ah this fails bootstrap because of `JULIA_CPU_TARGET="generic"`; I'll have to shuffle some code around.

maleadt commented:
Looking into the `range` test failure, it seems like parts of TwicePrecision are broken with this PR:

```julia
julia> x,y = Float16(0.5635), Float16(0.6133)
(Float16(0.5635), Float16(0.6133))

julia> Base.mul12(x,y)
(Float16(0.3457), 0.0)
```

Whereas before:

```julia
julia> x,y = Float16(0.5635), Float16(0.6133)
(Float16(0.5635), Float16(0.6133))

julia> Base.mul12(x,y)
(Float16(0.3455), Float16(0.0001106))
```

... which is more correct (`BigFloat(x) * BigFloat(y) = 0.345569610595703125`). The error seems to originate from `canonicalize2`:

```julia
julia> julia> fma(x,y,-(x*y))
Float16(0.0001106)

# before
julia> Base.canonicalize2(x*y, Float16(0.0001106))
(Float16(0.3455), Float16(0.0001106))

# after
julia> Base.canonicalize2(x*y, Float16(0.0001106))
(Float16(0.3455), 0.0)

# after, but in a function (changing intermediate precision)
julia> f(x,y) = Base.canonicalize2(x*y, Float16(0.0001106))
f (generic function with 2 methods)

julia> f(x,y)
(Float16(0.3457), 0.0)
```

I need to look at this some more, but @timholy maybe you have some thoughts already, as you wrote those utilities & range tests.

yuyichao commented:
Note that other than multiversioning issue, this is really going to throw off the GC root analysis. We assume normal llvm intrinsics do not call managed code and this can easily break that (though it'll certainly be fine most of the time.). It doesn't seem that any of these are too complex for C implementation (and I assume they usually are implemented in C anyway).

maleadt committed:
Port the Float16 runtime functions to C.

The Julia runtime wasn't safe wrt. GC operations.
I left the commit in for archival purposes, in case
we want to revisit this again later.

yuyichao commented on src/intrinsics.cpp:
```diff
@@ -1301,3 +1299,198 @@ static Value *emit_untyped_intrinsic(jl_codectx_t &ctx, intrinsic f, Value **arg
+        ret = sign | exp | sig;
+    }
+
+    float fret = *((float *)(&ret));
```
Use `memcpy` or mention that this requires `-fno-strict-aliasing`.

yuyichao commented on src/intrinsics.cpp:
```diff
@@ -1301,3 +1299,198 @@ static Value *emit_untyped_intrinsic(jl_codectx_t &ctx, intrinsic f, Value **arg
+
+static inline uint16_t float_to_half(float param)
+{
+    uint32_t f = *((uint32_t *)(&param));
```
Same here.

maleadt commented:
Ported the Float16 software intrinsics to C per @yuyichao's request.

The remaining failures are in the `ranges` test suite, caused by the twice-precision arithmetic breaking down as Float16 calculations are actually performed in Float32 (so `canonicalize2` and friends don't really work anymore). I couldn't find a way to convince LLVM to preserve 16-bit accuracy, or a way to get the current implementation working (except for forcing everything `@noinline` as values are returned in 16-bit registers, but that isn't an option of course). Maybe @timholy or @simonbyrne have some thoughts? FWIW, the range mentioned at the top of `twiceprecision.jl` works, `(Float16(0.1):Float16(0.1):Float16(0.3))[3] == Float16(0.3)`.

maleadt commented:
@nanosoldier `runtests(ALL, vs = ":master")`

maleadt commented:
Looking a bit closer at the twice precision code used by ranges: https://github.com/JuliaLang/julia/blob/8987f7ac6d284cb512d336279a74b5414c635757/base/twiceprecision.jl#L317
So IIUC Float16 doesn't even use TwicePrecision (all of the constructors I tested indeed used Float64 to increase precision), so I've disabled the (now) broken tests of those utilities on Float16 values.

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37510#issuecomment-704723630) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/60fdbb3ba57ab55213e18d45ad3542c67f19b77a/pkgeval/by_hash/1c84fa8_vs_7d3dac4/report.md). cc @maleadt

maleadt commented:
@nanosoldier `runtests(["BitIntegers", "BoltzmannMachines", "DIVAnd", "FastTransforms", "Graph500", "ImageFeatures", "IntervalTrees", "KrylovKit", "MIToS", "OptimKit", "QuadGK", "RandomizedPropertyTest", "Revise", "Sherlogs", "ThreadPools", "YAActL"], vs = ":master")`

maleadt commented:
@nanosoldier `runtests(["BitIntegers", "BoltzmannMachines", "DIVAnd", "FastTransforms", "Graph500", "ImageFeatures", "IntervalTrees", "MIToS", "QuadGK", "RandomizedPropertyTest", "Revise", "Sherlogs", "ThreadPools", "YAActL"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37510#issuecomment-704937170) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/2c54f6475536a70a1d7775f173267f74d269e530/pkgeval/by_hash/c8d800c_vs_8987f7a/report.md). cc @maleadt

maleadt commented:
Looking into test failures:
- BitIntegers: has special casing for Float16 that can just be removed. Relied on the `Float16(x::Integer)` we used to have, where for other floating-point types we only have conversions to Base's integer types. Put back or fix the package?
- FastTransforms: legit issue, reduced to `sinpi(Float16(-0.5))` returning different results.

simonbyrne commented:
So LLVM uses Float32 intermediate precision? I wonder what Swift does here, since they now offer Float16 support as well.

If that is the case, I think the easiest solution would be to promote to Float32 wherever we use these extended precision tricks.

simonbyrne commented:
Alternatively we could convert to Float32 and back again for all functions (which will give correctly rounded results for all basic arithmetic ops except fma). I'll try to explore this a bit tonight.

vchuravy commented:
Looking at the PRs that added `Float16` they generally seem to promote to Float32 for Float16. (https://github.com/apple/swift-numerics/pull/136). CUDA has a similar strategy in that they use faster less precise Float32 routines and then convert to Float16 + fix some of the bit patterns. I haven't seen anyone really attempting to define Float16 math generically.

But yes LLVM performs Float16 operations in extended precision (since there are no Float16 registers), and if my understanding is correct that is indeed legal...

simonbyrne commented:
> But yes LLVM performs Float16 operations in extended precision (since there are no Float16 registers), and if my understanding is correct that is indeed legal...

I'm not sure what you mean by legal here: we currently define all operations to return results of the same type as their inputs. Switching to allowing higher intermediate precision would be a significant breaking change as it fundamentally breaks the semantics of the language (as this PR shows).

Can you provide any links on how CUDA handles this? If there is no way to obtain performant code on GPUs without weakening these guarantees then its worth considering, but it is worth noting that Swift explicitly also provides the same strict behaviour.

vchuravy commented:
So I just looked at what Clang does:

https://github.com/llvm/llvm-project/blob/a5ef2e0a1e3bf5a76b0c170ca7791564ab1a8375/clang/lib/CodeGen/CodeGenTypes.cpp#L491-L512

https://github.com/llvm/llvm-project/blob/a5ef2e0a1e3bf5a76b0c170ca7791564ab1a8375/clang/lib/CodeGen/CodeGenTypes.cpp#L298-L303

- https://reviews.llvm.org/D33719
- https://reviews.llvm.org/D57188

IIUC they define a storage variant  `__fp16` and a numerical variant `_Float16`. The numerical variant is defined by  ISO/IEC TS 18661-3:2015.

`_Float16` ABI is not yet defined by many platforms outside ARM it seems.

> ``_Float16`` arithmetic will be performed using native half-precision support
when available on the target (e.g. on ARMv8.2a); otherwise it will be performed
at a higher precision (currently always ``float``) and then truncated down to
``_Float16``.  Note that C and C++ allow intermediate floating-point operands
of an expression to be computed with greater precision than is expressible in
their type, so Clang may avoid intermediate truncations in certain cases; this may
lead to results that are inconsistent with native arithmetic.

>  but it is worth noting that Swift explicitly also provides the same strict behaviour.

I haven't found the place in Swift where they are enforcing the native semantics on platforms that don't have native Float16. I only found that in `swift-numerics` they [promote to Float32 for library functions](https://github.com/apple/swift-numerics/blob/main/Sources/RealModule/Float16%2BReal.swift), which I assume is because they don't have implementations of those library functions.  I suspect that if you were to take our implementation and rewrite it in Swift you would get the same behavior.

maleadt commented:
> I suspect that if you were to take our implementation and rewrite it in Swift you would get the same behavior.

Correct:

```swift
func foo(x: Float16) -> Float16 {
  let s: Float16 = 1024
  let t: Float16 = 3*s
  let rx: Float16 = x-((x+t)-t)
  return rx
}

print(foo(x: 0.5))
```

```
0.0
```

Let's outline the addition:

```swift
func bar(x: Float16) -> Float16 {
  let s: Float16 = 1024
  let t: Float16 = 3*s
  let rx: Float16 = x-(bar_plus(a:x, b:t)-t)
  return rx
}

@inline(never) func bar_plus(a: Float16, b: Float16) -> Float16 {
  return a+b
}

print(bar(x: 0.5))
```

```
0.5
```

maleadt commented:
GCC has `-ffloat-store`:

> Do not store floating point variables in registers, and inhibit other options that might change whether a floating point value is taken from a register or memory.
>
> This option prevents undesirable excess precision on machines such as the 68000 where the floating registers (of the 68881) keep more precision than a double is supposed to have. Similarly for the x86 architecture. For most programs, the excess precision does only good, but a few programs rely on the precise definition of IEEE floating point. Use -ffloat-store for such programs, after modifying them to store all pertinent intermediate computations into variables. 
```

maleadt commented:
Yeah the current semantics are not OK. Reduced from the RandomizedPropertyTest failure:

```julia
function generate(rng, T::Type, z₀, r)
  z = Complex{T}(Inf, Inf)
  while !(abs(z - z₀) < r)
    z = r * complex(2rand(rng, T)-1, 2rand(rng, T)-1) + z₀
  end
  return z
end

function main(n)
    rng = MersenneTwister(0)
    r = 5
    z₀ = 4 + 2im
    z = nothing
    for i in 1:n
        z = generate(rng, Float16, z₀, r)
    end
    z
end
```

Depending on whether the type argument to `generate` is like here, or specialized (`::Type{T} where T`), the function is inlined and more operations fall under extended precision, resulting in different values.

The actual test failure is slightly different, where random numbers are generated like here to pass a certain condition (`abs(z - z₀) < r`), but after a while a Float16 number is generated where the condition holds under Float32 extended precision, but not anymore after returning through memory and getting truncated to an actual Float16 value.

I had already worked through the other test failures, fixing up some operations in Base to be compatible with these semantics. Will push those commits.

vtjnash commented:
> GCC has -ffloat-store: 

Yes, julia has that too. It's called `JL_NEED_FLOATTEMP_VAR`, but currently only enabled on `_CPU_X86_`.

simonbyrne commented:
@stephentyrone: is Swift's handling of intermediate precision of arithmetic operations (in particular for `Float16`) described anywhere?

simonbyrne commented:
Unfortunately it's always going to be a tradeoff: allowing higher intermediate precision can give you better performance on platforms where `Float16` isn't supported natively, as well as reducing some numerical errors, but at the cost of some non-determinism (and differences between platforms).

I'm happy to go with the Clang behaviour here, as long as we document it.

oscardssmith commented:
I think this is reasonable. Extended precision tricks in fp16 are probably always a really dumb idea. Float32 will basically always be quicker and more precise.

stephentyrone commented:
@simonbyrne Swift doesn't document Float16 specifically, but as a matter of policy we never carry excess precision in floating-point. That leaves some optimizations on the table (and can give less accurate results), but helps with portability and predictability.

vchuravy commented:
@stephentyrone Thanks for chiming in. I think those are the semantics we ultimately want for ourselves. https://github.com/JuliaLang/julia/pull/37510#issuecomment-705359640 has an example in Swift on x86_64 that shows operations being performed in extended precision. It would be great if we could  fix this in LLVM, but I suspect that it will require changes in the backend/legalizer for x86. Clang disabled `_Float16` on non-ARM targets since the ABI is not yet defined, and explicitly states (https://clang.llvm.org/docs/LanguageExtensions.html#half-precision-floating-point) that it can perform operations on `_Float16` in in extended precision. So if Swift & Julia want truncating semantics we will either have to implement that in the frontend or get support for that in the backend.

maleadt committed:
Explicitly demote Float16 operations to Float32.

maleadt commented:
OK, trying a different approach, with a late LLVM pass to manually extend and truncate operations on `Float16`s to their `Float32` equivalents to get the same semantics we had when using `Int16`. This seems promising (albeit at a cost, of course), passing almost all tests right of the bat. Let's see if that holds for packages too:

@nanosoldier `runtests(ALL, vs = ":master")`

TODO:
- there's a codegen bug with neg(NaN16)
- runtime intrinsics

oscardssmith commented:
> OK, trying a different approach, with a late LLVM pass to manually extend and truncate operations on `Float16`s to their `Float32` equivalents to get the same semantics we had when using `Int16`. 
This seems like a really good approach. It would also make custom Float16 special functions potentially worth implementing. They could be much faster than current, by using Float32 for internal calculations, but using smaller polynomials to only get .5ulps Float16. The only tricky bit is that any method using Float16 probably needs guaranteed accuracy to the .5ULP rounding the wrong way is probably unacceptable error the good news is that testing every possible pair of Float16 can be done in ~1 minute.

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37510#issuecomment-707946069) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/9b00c5797ffdf03221bcf55fdddfb950bfa2da52/pkgeval/by_hash/3d99aeb_vs_a1da84c/report.md). cc @maleadt

vchuravy commented on src/llvm-demote-float16.cpp:
```diff
@@ -0,0 +1,115 @@
+
+Pass *createDemoteFloat16Pass()
+{
+    return new DemoteFloat16Pass();
```
Should have a C-API constructor as well

maleadt commented:
@nanosoldier `runtests(["Autologistic", "Circuitscape", "ClusterManagers", "ControlSystemIdentification", "Convex1d", "DataDrivenDiffEq", "Graph500", "KissMCMC", "MPSKit", "Revise", "Transducers", "YAActL"], vs = ":master")`

maleadt committed:
Use LLVM's FNeg to get a compliant neg_float for all Float types.

maleadt committed:
Implement Float16 runtime intrinsics.

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37510#issuecomment-708169408) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/c7ba8280199654aed73ed629a46cc9e44096b1e9/pkgeval/by_hash/3d99aeb_vs_a1da84c/report.md). cc @maleadt

maleadt commented:
This should be complete now, and passes all tests locally. Let's do another PkgEval run:

@nanosoldier `runtests(ALL, vs = ":master")`

maleadt commented on src/runtime_intrinsics.c:
```diff
@@ -834,15 +902,17 @@ cvt_iintrinsic(LLVMFPtoUI, fptoui)
 #define fpext(pr, a) \
-        if (!(osize > 8 * sizeof(a))) \
-            jl_error("fpext: output bitsize must be > input bitsize"); \
+        if (!(osize >= 8 * sizeof(a))) \
```
This is the only behavioral change, now allowing no-op `fpext` to simplify the implementation of Float16 intrinsics. Shouldn't matter in practice.

vtjnash replied:
It’s invalid for llvm IR, so I’d just given the same restriction here

maleadt commented:
And FWIW, moving the pass that truncates/extends Float16 operations earlier in the pipeline (to let LLVM optimize) gives the following for `mul12`:

```diff
@@ -1,15 +1,9 @@
 julia> @code_llvm Base.mul12(x,y)
 define [2 x half] @julia_mul12(half %0, half %1) #0 {
 top:
-    %0 = fpext half %0 to float
-    %1 = fpext half %1 to float
-    %2 = fmul float %0, %1
-    %2 = fptrunc float %2 to half
+    %2 = fmul half %0, %1
     %3 = fcmp une half %2, 0xH0000
-    %2 = fpext half %2 to float
-    %2 = fpext half %2 to float
-    %4 = fsub float %2, %2
-    %4 = fptrunc float %4 to half
+    %4 = fsub half %2, %2
     %5 = fcmp oeq half %4, 0xH0000
     %6 = fneg half %2
     %7 = fpext half %0 to float
@@ -17,18 +11,9 @@
     %9 = fpext half %6 to float
     %10 = call float @llvm.fma.f32(float %7, float %8, float %9)
     %11 = fptrunc float %10 to half
-    %2 = fpext half %2 to float
-    %11 = fpext half %11 to float
-    %12 = fadd float %2, %11
-    %12 = fptrunc float %12 to half
-    %2 = fpext half %2 to float
-    %12 = fpext half %12 to float
-    %13 = fsub float %2, %12
-    %13 = fptrunc float %13 to half
-    %13 = fpext half %13 to float
-    %11 = fpext half %11 to float
-    %14 = fadd float %13, %11
-    %14 = fptrunc float %14 to half
+    %12 = fadd half %2, %11
+    %13 = fsub half %2, %12
+    %14 = fadd half %13, %11
     %15 = and i1 %3, %5
     %.fca.0.insert7 = insertvalue [2 x half] undef, half %2, 0
```

Cleans up a lot, and keeps the `ranges` TwicePrecision tests working, but I don't think this is legal wrt. our Float16 semantics. For example, the initial `mul` + `cmp` might behave differently under Float32 extended precision. I think I'll just add a GVN pass to clean up identical conversions (instcombine is too aggressive, and is responsible for most of the changes seen above).

vchuravy reviewed (approved):
This is looking great! Fantastic work Tim.

Should we try turning on native Float16 support on AArch64/ARM?

In https://reviews.llvm.org/D57188 they are turned on for all AARCH64 targets.

>  // All AArch64 implementations support ARMv8 FP, which makes half a legal type.
>  HasLegalHalfType = true;
> HasFloat16 = true;

CC: @yuyichao

yuyichao commented:
AFAICT `half` is a native type on aarch64 though arithmetics on them is an optional feature. This is why the ccall ABI for aarch64 already uses the native half type IIRC. I don't know what's the desired property here.

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37510#issuecomment-708354377) has completed - possible new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/6c273edbe998059dcfb9f44b619a54c6938e6c62/pkgeval/by_hash/2c77780_vs_15e1d59/report.md). cc @maleadt

maleadt commented:
@nanosoldier `runtests(["DynamicalBilliards", "ExcelReaders", "FameSVD", "FlashWeave", "Genie", "ITensors", "IntervalTrees", "LoopThrottle", "Manifolds", "MemPool", "OptimKit", "Pidfile"], vs = ":master")`

nanosoldier commented:
[Your package evaluation job](https://github.com/JuliaLang/julia/pull/37510#issuecomment-708523687) has completed - no new issues were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/02107c0b96d6c8a04288ae3a3ac499ea72527196/pkgeval/by_hash/b1b4def_vs_d60e2da/report.md). cc @maleadt

vchuravy commented:
> Happy to take a look at enabling native half on ARM, but I'd prefer to get this merged first before the 1.6 branch.

Merging this first is okay by me. I am excited to see this land.

maleadt commented:
CI and PkgEval is clean, so let's merge this! Happy to do some follow-up development if anybody would still review post-merge, but I'd like to get this in before the 1.6 branch.
